### PR TITLE
Add tutorial on hyperparameter tuning with HParams

### PR DESCRIPTION
**Pull request by GalOshri**
Add a tutorial doc for the HParams dashboard.

This tutorial shows how to adapt a TF experiment to log the right summaries for HParams and also demonstrates the dashboard within TensorBoard.

### CODE REVIEW

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+    },
+    {
+      "metadata": {
+        "id": "Z3Q5v28XaCt8",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "The HParams dashboard has three different views, with various useful information:\n",
+        "* The **Table View** lists the runs, their hyperparameters, and their metrics.\n",
```
Comment (erzel) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Add that you can click on a table row / par-coords line / scatter-plot marker and see a plot of the metrics as a function of training step for that session. I know that in this tutorial we have just one step, so the line won't be very interesting, but we can point to that capability which I think many users will find valuable.

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+      ]
+    },
+    {
+      "metadata": {
+        "id": "fh3p0DtcdOK1",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "Download a set of logs with more experiments:"
```
Comment (erzel) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Nice idea of making the tutorial lightweight and providing downloadable summaries.

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb [hunk 3]
```diff
@@ -0,0 +1,599 @@
+{
```
Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>Update copyright. (The original notebook template needs to be fixed.)</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>Consider linking the first use of "hyperparameters" to https://en.wikipedia.org/wiki/Hyperparameter_(machine_learning)</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>nit: for clarity, I suggest separating these into:</p><p><br></p><p>default python imports</p><p><br></p><p>tensorflow imports</p><p><br></p><p>other imports.</p><p><br></p><p>So:</p><p><br></p><p>import datetime</p><p><br></p><p>import tensorflow as tf</p><p><strong>from</strong> <strong>tensorboard.plugins.hparams</strong> <strong>import</strong> api_pb2</p><p><strong>from</strong> <strong>tensorboard.plugins.hparams</strong> <strong>import</strong> summary <strong>as</strong> hparams_summary</p><p><br></p><p><strong>from</strong> <strong>google.protobuf</strong> <strong>import</strong> struct_pb2</p><p><br></p><p><br></p><p><br></p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>Should you clear outputs?</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>Consider rewording slightly to:</p><p><br></p><p>Log an experiment summary. This is how the HParams dashboard knows what the hyperparameters and metrics are.</p><p><br></p><p><br></p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>I don't see the run_dir parameter used here. Please remove.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>Accuracy should be passed to the metrics parameter of model.compile().</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>This should be prefaced by something like:</p><p><br></p><p>"To further explore the capabilities of the HParams plugin, download a set of pregenerated logs ..."</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>I feel like a motion gif or short video would be ideal here.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+      "cell_type": "markdown",
+      "source": [
+        "When building machine learning models, you need to choose various hyperparameters, such as the dropout rate in a layer or the learning rate. These decisions impact model metrics, such as accuracy. Therefore, an important step in the machine learning workflow is to identify the best hyperparameters for your problem, which often involves experimentation. This process is known as \"Hyperparameter Optimization\" or \"Hyperparameter Tuning\".\n",
+        "\n",
+        "The HParams dashboard in TensorBoard provides several tools to help with this process of identifying the best experiment or which sets of hyperparameters are most promising. \n",
+        "\n",
+        "This tutorial will focus on the following steps:\n",
+        "1. Experiment setup and HParams summary\n",
+        "2. Adapt TensorFlow runs to log hyperparameters and metrics\n",
+        "3. Start runs and log them in one directory\n",
```
[the remark this answers is not in the document]
Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Will change to "one parent directory". Let me know if there is a better description.

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "DCbssXK_Wle3",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "Log an experiment summary so the HParams dashboard knows what are the hyperparameters and metrics:"
```
[the remark this answers is not in the document]
Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Thanks for the pointer.

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+    },
+    {
+      "metadata": {
+        "id": "hG-zalNfW5Zl",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "def train_test_model(run_dir, hparams):\n",
```
[the remark this answers is not in the document]
Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Removed.

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+    {
+      "metadata": {
+        "id": "Z3Q5v28XaCt8",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "The HParams dashboard has three different views, with various useful information:\n",
+        "* The **Table View** lists the runs, their hyperparameters, and their metrics.\n",
+        "* The **Parallel Coordinates View** shows each run as a line going through an axis for each hyperparemeter and metric. Click and drag on any axis to highlight only sessions that pass through it. This can be useful for identifying which groups of hyperparameters are most important.\n",
```
[the remark this answers is not in the document]
Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Good suggestion, updated.

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+    },
+    {
+      "metadata": {
+        "id": "4RPGbR9EWd4o",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "The left pane of the dashboard provides filtering capabilities:\n",
+        "- Filter which hyperparameters/metrics are shown in the dashboard\n",
```
[the remark this answers is not in the document]
Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Added.

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb (hunk 1 above)
Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Added.

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb (hunk 2 above)
Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Thanks! It's better than waiting 30 minutes for the tutorial to run :).

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb (hunk 3 above)
Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Updated.

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Good idea. Added.

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Updated.

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: This is necessary for now given the discussion [here](https://github.com/tensorflow/tensorboard/pull/1909#discussion_r260850621)

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Reworded.

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Removed.

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: What do you mean by this? Accuracy is used as the metric in `model.compile()`

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Added.

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Adding an image (or GIF or video) would be a short-term solution until the docs site is updated. Something with motion might make the experience even more confusing when running within Colab (there will be a live TensorBoard and then something moving under it). I'm tempted to leave this as it is for now. WDYT?

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+      "source": [
+        "def run(run_dir, hparams):\n",
+        "  writer = tf.summary.create_file_writer(run_dir)\n",
+        "  summary_start = hparams_summary.session_start_pb(hparams=hparams)\n",
+        "\n",
+        "  with writer.as_default():\n",
+        "    accuracy = train_test_model(run_dir, hparams)\n",
+        "    summary_end = hparams_summary.session_end_pb(api_pb2.STATUS_SUCCESS)\n",
+        "      \n",
+        "    tf.summary.scalar('accuracy', accuracy, step=0, description=\"The accuracy\")\n",
```
[the remark this answers is not in the document]
Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Good idea. Addressed.

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb [hunk 10]
```diff
@@ -0,0 +1,603 @@
+{
```
Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>'the best experiment or most promising sets of hyperparameters."</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>Missing close parens.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>provide -&gt; "you'll provide the training function"</p><p><br></p><p>Also, the second sentence is really long for technical doc. Consider breaking it up.</p><p><br></p><p>The training code will look familiar, except you won't hardcode the hyperparameters. Instead, you'll provide a hparams dictionary to the training function ... for a particular training run. You'll then return the model's accuracy.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p># For clarity</p><p>_, accuracy = model.evaluate()</p><p>return accuracy</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (manivaradarajan) [inline docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>The first phrase is strange. Consider rewording to:</p><p><br></p><p>"You now have the ability to train with different sets of hyperparameters. You'll now try all combinations ..."</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/tensorboard/pull/1926/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb (hunk 3 above)
Comment (manivaradarajan) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>Yeah, simplicity is better.</p>

Comment (manivaradarajan) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: <p>My mistake</p>

File: docs/r2/hyperparameter_tuning_with_hparams.ipynb (hunk 10 above)
Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Addressed.

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Thanks for catching this.

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Addressed.

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Indeed, saying "runs" three times in a sentence is not ideal.

Comment (GalOshri) [reply docs/r2/hyperparameter_tuning_with_hparams.ipynb]: Updated

**Comment by lamberta** [comment]
Make sure to add it to the [_book.yaml](https://github.com/tensorflow/tensorboard/blob/master/docs/_book.yaml) TOC :)

**Comment by GalOshri** [comment]
> Make sure to add it to the [_book.yaml](https://github.com/tensorflow/tensorboard/blob/master/docs/_book.yaml) TOC :)

Thanks for the reminder! Done.
